contracts/test/v0.8/PermissionedForwardProxy.ts

```diff
@@ -0,0 +1,153 @@
+            .setPermission(
+              await personas.Carol.getAddress(),
+              await personas.Eddy.getAddress(),
+            ),
+        ).to.be.revertedWith('Only callable by owner')
+      })
+    })
+
+    describe('when called by the owner', () => {
+      it('adds the permission to the proxy', async () => {
+        await controller
+          .connect(personas.Carol)
+          .setPermission(
+            await personas.Carol.getAddress(),
+            await personas.Eddy.getAddress(),
+          )
+
+        expect(
+          await controller.getPermission(await personas.Carol.getAddress()),
+        ).to.be.equal(await personas.Eddy.getAddress())
+      })
+    })
+  })
+
+  describe('#removePermission', () => {
+    beforeEach(async () => {
+      // Add permission before testing
+      await controller
+        .connect(personas.Carol)
+        .setPermission(
+          await personas.Carol.getAddress(),
+          await personas.Eddy.getAddress(),
+        )
+    })
+
+    describe('when called by a non-owner', () => {
+      it('reverts', async () => {
+        await expect(
+          controller
+            .connect(personas.Eddy)
+            .removePermission(await personas.Carol.getAddress()),
+        ).to.be.revertedWith('Only callable by owner')
+      })
+    })
+
+    describe('when called by the owner', () => {
+      it('removes the permission to the proxy', async () => {
+        await controller
+          .connect(personas.Carol)
+          .removePermission(await personas.Carol.getAddress())
+
+        expect(
+          await controller.getPermission(await personas.Carol.getAddress()),
+        ).to.be.equal(ethers.constants.AddressZero)
+      })
+    })
+  })
+
+  describe('#forward', () => {
+    describe('when permission does not exist', () => {
+      it('reverts', async () => {
+        await expect(
+          controller
+            .connect(personas.Carol)
+            .forward(await personas.Eddy.getAddress(), '0x'),
+        ).to.be.revertedWith(PERMISSION_NOT_SET)
+      })
+    })
+
+    describe('when permission exists', () => {
+      beforeEach(async () => {
+        // Add permission before testing
+        await controller
+          .connect(personas.Carol)
+          .setPermission(await personas.Carol.getAddress(), counter.address)
+      })
+
+      it('calls target successfully', async () => {
+        await controller
+          .connect(personas.Carol)
+          .forward(
+            counter.address,
+            counter.interface.encodeFunctionData('increment'),
+          )
+
+        expect(await counter.count()).to.be.equal(1)
+      })
+
+      it('reverts when target reverts', async () => {
+        await expect(
+          controller
+            .connect(personas.Carol)
+            .forward(
+              counter.address,
+              counter.interface.encodeFunctionData('alwaysRevert'),
+            ),
+        ).to.be.reverted
```

Comment:
Javascript VM is not able to parse the exact error since it's defined on another contract